File: RoadRunnerJS/src/commonMain/kotlin/trajectory/TrajectoryBuilder.kt
package trajectory

import geometry.Pose2d
import geometry.Vector2d
import path.Path
import path.PathBuilder
import profile.MotionState
import trajectory.constraints.TrajectoryAccelerationConstraint
import trajectory.constraints.TrajectoryVelocityConstraint
import util.Angle
import kotlin.js.JsExport
import kotlin.js.JsName
import kotlin.math.PI

private fun MotionState.zeroPosition() = MotionState(0.0, v, a, j)

private data class IntervalVelocityConstraint(
    val start: Double,
    val end: Double,
    val constraint: TrajectoryVelocityConstraint
)

private class PiecewiseVelocityConstraint(
    private val baseConstraint: TrajectoryVelocityConstraint,
    private val constraintOverrides: List<IntervalVelocityConstraint>
) : TrajectoryVelocityConstraint() {
    override fun get(s: Double, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): Double {
        for (interval in constraintOverrides) {
            val (start, end, constraint) = interval
            if (s in start..end) {
                return constraint[s, pose, deriv, baseRobotVel]
            }
        }
        return baseConstraint[s, pose, deriv, baseRobotVel]
    }
}

private data class IntervalAccelerationConstraint(
    val start: Double,
    val end: Double,
    val constraint: TrajectoryAccelerationConstraint
)

private class PiecewiseAccelerationConstraint(
    private val baseConstraint: TrajectoryAccelerationConstraint,
    private val constraintOverrides: List<IntervalAccelerationConstraint>
) : TrajectoryAccelerationConstraint() {
    override fun get(s: Double, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): Double {
        for (interval in constraintOverrides) {
            val (start, end, constraint) = interval
            if (s in start..end) {
                return constraint[s, pose, deriv, baseRobotVel]
            }
        }
        return baseConstraint[s, pose, deriv, baseRobotVel]
    }
}

/**
 * Builder for trajectories with *dynamic* constraints.
 */
@JsExport
class TrajectoryBuilder constructor(
    startPose: Pose2d?,
    startTangent: Double?,
    trajectory: Trajectory?,
    t: Double?,
    private val baseVelConstraint: TrajectoryVelocityConstraint,
    private val baseAccelConstraint: TrajectoryAccelerationConstraint,
    private val start: MotionState,
    private val resolution: Double
) {
    private var pathBuilder: PathBuilder = if (startPose == null) {
        PathBuilder(trajectory!!.path, trajectory.profile[t!!].x)
    } else {
        PathBuilder(startPose, startTangent!!)
    }

    /**
     * Create a builder from a start pose and motion state. This is the recommended constructor for creating
     * trajectories from rest.
     */
//    @JsName("TrajectoryBuilderConstructor2")
//    constructor(
//        startPose: Pose2d,
//        startTangent: Double = startPose.heading,
//        baseVelConstraint: TrajectoryVelocityConstraint,
//        baseAccelConstraint: TrajectoryAccelerationConstraint,
//        resolution: Double = 0.25
//    ) : this(
//        startPose,
//        startTangent,
//        null,
//        null,
//        baseVelConstraint,
//        baseAccelConstraint,
//        MotionState(0.0, 0.0, 0.0),
//        resolution
//    )
//
//    @JsName("TrajectoryBuilderConstructor3")
//    constructor(
//        startPose: Pose2d,
//        reversed: Boolean,
//        baseVelConstraint: TrajectoryVelocityConstraint,
//        baseAccelConstraint: TrajectoryAccelerationConstraint,
//        resolution: Double = 0.25
//    ) : this(
//        startPose,
//        Angle.norm(startPose.heading + if (reversed) PI else 0.0),
//        baseVelConstraint,
//        baseAccelConstraint,
//        resolution
//    )
//
//    /**
//     * Create a builder from an active trajectory. This is useful for interrupting a live trajectory and smoothly
//     * transitioning to a new one.
//     */
//    @JsName("TrajectoryBuilderConstructor4")
//    constructor(
//        trajectory: Trajectory,
//        t: Double,
//        baseVelConstraint: TrajectoryVelocityConstraint,
//        baseAccelConstraint: TrajectoryAccelerationConstraint,
//        resolution: Double = 0.25
//    ) : this(
//        null,
//        null,
//        trajectory,
//        t,
//        baseVelConstraint,
//        baseAccelConstraint,
//        trajectory.profile[t].zeroPosition(),
//        resolution
//    )

    private val velConstraintOverrides = mutableListOf<IntervalVelocityConstraint>()
    private val accelConstraintOverrides = mutableListOf<IntervalAccelerationConstraint>()

    private fun addSegment(
        add: () -> Unit,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ): TrajectoryBuilder {
        val start = pathBuilder.build().length()

        add()

        val end = pathBuilder.build().length()

        if (velConstraintOverride != null) {
            velConstraintOverrides.add(IntervalVelocityConstraint(start, end, velConstraintOverride))
        }

        if (accelConstraintOverride != null) {
            accelConstraintOverrides.add(IntervalAccelerationConstraint(start, end, accelConstraintOverride))
        }

        return this
    }

    /**
     * Adds a line segment with tangent heading interpolation.
     *
     * @param endPosition end position
     * @param constraintsOverride segment-specific constraints
     */
    fun lineTo(
        endPosition: Vector2d,
        velConstraintOverride: TrajectoryVelocityConstraint,
        accelConstraintOverride: TrajectoryAccelerationConstraint
    ) =
        addSegment({ pathBuilder.lineTo(endPosition) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a line segment with constant heading interpolation.
     *
     * @param endPosition end position
     * @param constraintsOverride segment-specific constraints
     */
    fun lineToConstantHeading(
        endPosition: Vector2d,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.lineToConstantHeading(endPosition) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a line segment with linear heading interpolation.
     *
     * @param endPose end pose
     * @param constraintsOverride segment-specific constraints
     */
    fun lineToLinearHeading(
        endPose: Pose2d,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.lineToLinearHeading(endPose) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a line segment with spline heading interpolation.
     *
     * @param endPose end pose
     * @param constraintsOverride segment-specific constraints
     */
    fun lineToSplineHeading(
        endPose: Pose2d,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.lineToSplineHeading(endPose) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a strafe path segment.
     *
     * @param endPosition end position
     * @param constraintsOverride segment-specific constraints
     */
    fun strafeTo(
        endPosition: Vector2d,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.strafeTo(endPosition) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a line straight forward.
     *
     * @param distance distance to travel forward
     * @param constraintsOverride segment-specific constraints
     */
    fun forward(
        distance: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.forward(distance) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a line straight backward.
     *
     * @param distance distance to travel backward
     * @param constraintsOverride segment-specific constraints
     */
    fun back(
        distance: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.back(distance) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a segment that strafes left in the robot reference frame.
     *
     * @param distance distance to strafe left
     * @param constraintsOverride segment-specific constraints
     */
    fun strafeLeft(
        distance: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.strafeLeft(distance) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a segment that strafes right in the robot reference frame.
     *
     * @param distance distance to strafe right
     * @param constraintsOverride segment-specific constraints
     */
    fun strafeRight(
        distance: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.strafeRight(distance) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a spline segment with tangent heading interpolation.
     *
     * @param endPosition end position
     * @param endTangent end tangent
     * @param constraintsOverride segment-specific constraints
     */
    fun splineTo(
        endPosition: Vector2d,
        endTangent: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.splineTo(endPosition, endTangent) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a spline segment with constant heading interpolation.
     *
     * @param endPosition end position
     * @param endTangent end tangent
     * @param constraintsOverride segment-specific constraints
     */
    fun splineToConstantHeading(
        endPosition: Vector2d,
        endTangent: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.splineToConstantHeading(endPosition, endTangent) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a spline segment with linear heading interpolation.
     *
     * @param endPose end pose
     * @param endTangent end tangent
     * @param constraintsOverride segment-specific constraints
     */
    fun splineToLinearHeading(
        endPose: Pose2d,
        endTangent: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.splineToLinearHeading(endPose, endTangent) }, velConstraintOverride, accelConstraintOverride)

    /**
     * Adds a spline segment with spline heading interpolation.
     *
     * @param endPose end pose
     * @param constraintsOverride segment-specific constraints
     */
    fun splineToSplineHeading(
        endPose: Pose2d,
        endTangent: Double,
        velConstraintOverride: TrajectoryVelocityConstraint?,
        accelConstraintOverride: TrajectoryAccelerationConstraint?
    ) =
        addSegment({ pathBuilder.splineToSplineHeading(endPose, endTangent) }, velConstraintOverride, accelConstraintOverride)

    fun buildTrajectory(
        path: Path,
        temporalMarkers: List<TemporalMarker>,
        displacementMarkers: List<DisplacementMarker>,
        spatialMarkers: List<SpatialMarker>
    ): Trajectory {
        val goal = MotionState(path.length(), 0.0, 0.0)
        return TrajectoryGenerator.generateTrajectory(
            path,
            PiecewiseVelocityConstraint(baseVelConstraint, velConstraintOverrides),
            PiecewiseAccelerationConstraint(baseAccelConstraint, accelConstraintOverrides),
            start,
            goal,
            temporalMarkers,
            displacementMarkers,
            spatialMarkers,
            resolution
        )
    }
}

File: src/roadrunnerjs/trajectory/TrajectoryBuilder.ts
import { MotionState } from "./../profile/MotionState";
import { norm } from "roadrunnerjs/util";
import { Pose2d, Vector2d } from "roadrunnerjs/geometry";
import { BaseTrajectoryBuilder } from "./BaseTrajectoryBuilder";
import { TrajectoryAccelerationConstraint } from "./constraints";
import { TrajectoryVelocityConstraint } from "./constraints/TrajectoryVelocityConstraint";
import { Trajectory } from "./Trajectory";
import { Path } from "roadrunnerjs/path";
import { TemporalMarker } from "./TemporalMarker";
import { DisplacementMarker } from "./DisplacementMarker";
import { SpatialMarker } from "./SpatialMarker";
import { generateTrajectory } from "./TrajectoryGenerator";

type IntervalVelocityConstraint = {
  start: number;
  end: number;
  constraint: TrajectoryVelocityConstraint;
};

class PiecewiseVelocityConstraint implements TrajectoryVelocityConstraint {
  constructor(
    private readonly baseConstraint: TrajectoryVelocityConstraint,
    private readonly constraintOverrides: IntervalVelocityConstraint[]
  ) {}

  get(s: number, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): number {
    for (const interval of this.constraintOverrides) {
      const { start, end, constraint } = interval;

      if (s > start && s < end) {
        return constraint.get(s, pose, deriv, baseRobotVel);
      }
    }
    return this.baseConstraint.get(s, pose, deriv, baseRobotVel);
  }
}

type IntervalAccelerationConstraint = {
  start: number;
  end: number;
  constraint: TrajectoryAccelerationConstraint;
};

class PiecewiseAccelerationConstraint
  implements TrajectoryAccelerationConstraint
{
  constructor(
    private readonly baseConstraint: TrajectoryAccelerationConstraint,
    private readonly constraintOverrides: IntervalAccelerationConstraint[]
  ) {}

  get(s: number, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d): number {
    for (const interval of this.constraintOverrides) {
      const { start, end, constraint } = interval;

      if (s > start && s < end) {
        return constraint.get(s, pose, deriv, baseRobotVel);
      }
    }
    return this.baseConstraint.get(s, pose, deriv, baseRobotVel);
  }
}

/**
 * Create a builder from a start pose and motion state. This is the recommended constructor for creating
 * trajectories from rest.
 */
type TrajectoryBuilderConstructor1 = {
  startPose: Pose2d;
  startTangent: number;
  baseVelConstraint: TrajectoryVelocityConstraint;
  baseAccelConstraint: TrajectoryAccelerationConstraint;
  resolution?: number;
};

type TrajectoryBuilderConstructor2 = {
  startPose: Pose2d;
  reversed: boolean;
  baseVelConstraint: TrajectoryVelocityConstraint;
  baseAccelConstraint: TrajectoryAccelerationConstraint;
  resolution?: number;
};

/**
 * Create a builder from an active trajectory. This is useful for interrupting a live trajectory and smoothly
 * transitioning to a new one.
 */
type TrajectoryBuilderConstructor3 = {
  trajectory: Trajectory;
  t: number;
  baseVelConstraint: TrajectoryVelocityConstraint;
  baseAccelConstraint: TrajectoryAccelerationConstraint;
  resolution?: number;
};

/**
 * Builder for trajectories with *dynamic* constraints.
 */
export class TrajectoryBuilder extends BaseTrajectoryBuilder<TrajectoryBuilder> {
  private readonly baseVelConstraint: TrajectoryVelocityConstraint;
  private readonly baseAccelConstraint: TrajectoryAccelerationConstraint;
  private readonly start: MotionState;
  private readonly resolution: number;

  private readonly velConstraintOverrides: IntervalVelocityConstraint[] = [];
  private readonly accelConstraintOverrides: IntervalAccelerationConstraint[] =
    [];

  constructor(
    p:
      | TrajectoryBuilderConstructor1
      | TrajectoryBuilderConstructor2
      | TrajectoryBuilderConstructor3
  ) {
    if ("startPose" in p) {
      if ("startTangent" in p) {
        super(p.startPose, p.startTangent, undefined, undefined);
      } else {
        super(
          p.startPose,
          norm(p.startPose.heading + (p.reversed ? Math.PI : 0)),
          undefined,
          undefined
        );
      }

      this.start = new MotionState(0, 0, 0);
    } else {
      super(undefined, undefined, p.trajectory, p.t);

      this.start = p.trajectory.profile.get(p.t).zeroPosition();
    }

    this.baseVelConstraint = p.baseVelConstraint;
    this.baseAccelConstraint = p.baseAccelConstraint;

    this.resolution = p.resolution ?? 0.25;
  }

  addSegment(
    add: () => void,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ): TrajectoryBuilder {
    const start = this.pathBuilder.build().length();

    add();

    const end = this.pathBuilder.build().length();

    if (velConstraintOverride != null) {
      this.velConstraintOverrides.push({
        start,
        end,
        constraint: velConstraintOverride,
      });
    }

    if (accelConstraintOverride != null) {
      this.accelConstraintOverrides.push({
        start,
        end,
        constraint: accelConstraintOverride,
      });
    }

    return this;
  }

  /**
   * Adds a line segment with tangent heading interpolation.
   *
   * @param endPosition end position
   * @param constraintsOverride segment-specific constraints
   */
  lineTo(
    endPosition: Vector2d,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.lineTo(endPosition),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a line segment with constant heading interpolation.
   *
   * @param endPosition end position
   * @param constraintsOverride segment-specific constraints
   */
  lineToConstantHeading(
    endPosition: Vector2d,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.lineToConstantHeading(endPosition),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a line segment with linear heading interpolation.
   *
   * @param endPose end pose
   * @param constraintsOverride segment-specific constraints
   */
  lineToLinearHeading(
    endPose: Pose2d,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.lineToLinearHeading(endPose),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a line segment with spline heading interpolation.
   *
   * @param endPose end pose
   * @param constraintsOverride segment-specific constraints
   */
  lineToSplineHeading(
    endPose: Pose2d,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.lineToSplineHeading(endPose),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a strafe path segment.
   *
   * @param endPosition end position
   * @param constraintsOverride segment-specific constraints
   */
  strafeTo(
    endPosition: Vector2d,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.strafeTo(endPosition),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a line straight forward.
   *
   * @param distance distance to travel forward
   * @param constraintsOverride segment-specific constraints
   */
  forward(
    distance: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.forward(distance),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a line straight backward.
   *
   * @param distance distance to travel backward
   * @param constraintsOverride segment-specific constraints
   */
  back(
    distance: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.back(distance),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a segment that strafes left in the robot reference frame.
   *
   * @param distance distance to strafe left
   * @param constraintsOverride segment-specific constraints
   */
  strafeLeft(
    distance: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.strafeLeft(distance),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a segment that strafes right in the robot reference frame.
   *
   * @param distance distance to strafe right
   * @param constraintsOverride segment-specific constraints
   */
  strafeRight(
    distance: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.strafeRight(distance),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a spline segment with tangent heading interpolation.
   *
   * @param endPosition end position
   * @param endTangent end tangent
   * @param constraintsOverride segment-specific constraints
   */
  splineTo(
    endPosition: Vector2d,
    endTangent: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.splineTo(endPosition, endTangent),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a spline segment with constant heading interpolation.
   *
   * @param endPosition end position
   * @param endTangent end tangent
   * @param constraintsOverride segment-specific constraints
   */
  splineToConstantHeading(
    endPosition: Vector2d,
    endTangent: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.splineToConstantHeading(endPosition, endTangent),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a spline segment with linear heading interpolation.
   *
   * @param endPose end pose
   * @param endTangent end tangent
   * @param constraintsOverride segment-specific constraints
   */
  splineToLinearHeading(
    endPose: Pose2d,
    endTangent: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.splineToLinearHeading(endPose, endTangent),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  /**
   * Adds a spline segment with spline heading interpolation.
   *
   * @param endPose end pose
   * @param constraintsOverride segment-specific constraints
   */
  splineToSplineHeading(
    endPose: Pose2d,
    endTangent: number,
    velConstraintOverride?: TrajectoryVelocityConstraint,
    accelConstraintOverride?: TrajectoryAccelerationConstraint
  ) {
    return this.addSegment(
      () => super.splineToSplineHeading(endPose, endTangent),
      velConstraintOverride,
      accelConstraintOverride
    );
  }

  buildTrajectory(
    path: Path,
    temporalMarkers: TemporalMarker[],
    displacementMarkers: DisplacementMarker[],
    spatialMarkers: SpatialMarker[]
  ): Trajectory {
    const goal = new MotionState(path.length(), 0.0, 0.0);
    return generateTrajectory(
      path,
      new PiecewiseVelocityConstraint(
        this.baseVelConstraint,
        this.velConstraintOverrides
      ),
      new PiecewiseAccelerationConstraint(
        this.baseAccelConstraint,
        this.accelConstraintOverrides
      ),
      this.start,
      goal,
      temporalMarkers,
      displacementMarkers,
      spatialMarkers,
      this.resolution
    );
  }
}
